<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = ''
USER = ''
PASSWORD = ''
HOST = ''
PORT = ''

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select 
                    count(id)
                from 
                    vacancies
            '''

In [5]:
# результат запроса
df_vac_3_1 = pd.read_sql_query(query_3_1, connection)
df_vac_3_1

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\245319810.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_3_1 = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select 
                    count(id)
                from 
                    employers
            '''

In [7]:
# результат запроса
df_emp_3_2 = pd.read_sql_query(query_3_2, connection)
df_emp_3_2

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\465477527.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select 
                    count(id)
                from 
                    areas
            '''

In [9]:
# результат запроса
df_areas_3_3 = pd.read_sql_query(query_3_3, connection)
df_areas_3_3

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3904955400.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas_3_3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select 
                    count(id)
                from 
                    industries
            '''

In [11]:
# результат запроса
df_ind_3_4 = pd.read_sql_query(query_3_4, connection)
df_ind_3_4

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\4153160748.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_ind_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

## Вывод
#### Согласно предварительному анализу, база данных содержит информацию немного менее чем о 50 тысяч вакансиях, размещенных более чем 23 тысячами работодателей из более чем 1 тысячи 300 регионов по почти 300 сферам деятельности.

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''select 
                    count(v.id) cnt,
                    a.name area
                from 
                    vacancies v
                join 
                    areas a on v.area_id = a.id
                group by 
                    area
                order by 
                    cnt desc
            '''

In [13]:
# результат запроса
df_vac_4_1 = pd.read_sql_query(query_4_1, connection)
df_vac_4_1

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\1487764822.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_4_1 = pd.read_sql_query(query_4_1, connection)


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''select 
                    count(id) cnt
                from 
                    vacancies
                where 
                    salary_from is not null or salary_to is not null
            '''

In [15]:
# результат запроса
df_vac_4_2 = pd.read_sql_query(query_4_2, connection)
df_vac_4_2

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\4089058662.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_4_2 = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''select 
                    round(avg(salary_from), 0) avg_salary_from,
                    round(avg(salary_to), 0) avg_salary_to
                from 
                    vacancies
            '''

In [17]:
# результат запроса
df_vac_4_3 = pd.read_sql_query(query_4_3, connection)
df_vac_4_3

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\1233760276.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_4_3 = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''select 
                    count(id) cnt,
                    schedule,
                    employment
                from 
                    vacancies
                group by 
                    schedule, employment
                order by 
                    cnt desc
            '''

In [19]:
# результат запроса
df_vac_4_4 = pd.read_sql_query(query_4_4, connection)
df_vac_4_4

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\1627134213.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_4_4 = pd.read_sql_query(query_4_4, connection)


cnt          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''select 
                    count(id) cnt,
                    experience
                from 
                    vacancies
                group by 
                    experience
                order by 
                    cnt asc
            '''

In [21]:
# результат запроса
df_vac_4_5 = pd.read_sql_query(query_4_5, connection)
df_vac_4_5

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\1657687306.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_4_5 = pd.read_sql_query(query_4_5, connection)


cnt          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

## Вывод
#### База данных содержит вакансии стран СНГ. Основная часть вакансий (25%) представлена в следующих городах (более 2 тысяч вакансий в каждом): Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Почти в половине представленных вакансий отражен ориентир по заработной плате. В целом по базе данных вакансии предлагаются с зарпатой в среднем от 71 до 111 тысяч рублей. Наиболее распростарненный график работы для соискателей - полная занятость. Тип рабочего графика в абсолютном большинстве - полный день, значительно реже редоагается удаленная работа. Работодатели рассматривают кандидатов с различным опытом работы, однако самыми востребованным являются специалисты с опытом от 1 года до 6 лет, порядка 15% вакансий для начинающих работников, и всего 2% вакансий для опытных соискателей с остажем от 6 лет. 

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''(select 
                    e.name,
                    count(v.id)
                from 
                    employers e
                join 
                    vacancies v on e.id = v.employer_id
                group by 
                    e.name
                order by 
                    count(v.id) desc
                -- берем первое место
                limit 1
                )
                union
                -- соединяем
                (select 
                    e.name,
                    count(v.id)
                from 
                    employers e
                join 
                    vacancies v on e.id = v.employer_id
                group by 
                    e.name
                order by 
                    count(v.id) desc
                -- берем четвертое место
                offset 4 limit 1
                )
                order by 
                    2 desc
            '''

In [23]:
# результат запроса
df_emp_5_1 = pd.read_sql_query(query_5_1, connection)
df_emp_5_1

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3976741550.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp_5_1 = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''select 
                    a.name,
                    count(e.id) employers_cnt,
                    count(v.id) vacancies_cnt
                from 
                    areas a
                left join 
                    employers e on e.area = a.id
                left join 
                    vacancies v on v.area_id = a.id
                group by 
                    a.name
                order by 
                    vacancies_cnt asc, employers_cnt desc
                limit 1
            '''

In [25]:
# результат запроса
df_area_5_2 = pd.read_sql_query(query_5_2, connection)
df_area_5_2

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\1275627807.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_area_5_2 = pd.read_sql_query(query_5_2, connection)


name  employers_cnt  vacancies_cnt
0  Россия            410              0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''select 
                    e.name,
                    count(distinct v.area_id) areas_cnt
                from 
                    employers e
                join 
                    vacancies v on e.id = v.employer_id
                group by 
                    e.name
                order by 
                    areas_cnt desc
            '''

In [27]:
# результат запроса
df_emp_5_3 = pd.read_sql_query(query_5_3, connection)
df_emp_5_3

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\840636785.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp_5_3 = pd.read_sql_query(query_5_3, connection)


name  areas_cnt
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
14761                  UniSol          1
14762            UNISTORY LLC          1
14763                   UNIT6          1
14764     United Distribution          1
14765              UNITEL LLC          1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''select 
                    count(e.id)
                from 
                    employers e
                left join 
                    employers_industries i on e.id = i.employer_id
                where 
                    i.industry_id is null
            '''

In [29]:
# результат запроса
df_emp_5_4 = pd.read_sql_query(query_5_4, connection)
df_emp_5_4

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\878668837.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
industry_n = 4 # количество сфер деятельности

# текст запроса
query_5_5 = f'''select 
                    e.name,
                    count(i.industry_id) industry_cnt
                from 
                    employers e
                left 
                    join employers_industries i on e.id = i.employer_id
                group 
                    by e.name
                having 
                    count(i.industry_id) = {industry_n}
                order by 
                    e.name
                offset 2 limit 1
            '''

In [33]:
# результат запроса
df_emp_5_5 = pd.read_sql_query(query_5_5, connection)
df_emp_5_5

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3480483538.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp_5_5 = pd.read_sql_query(query_5_5, connection)


name  industry_cnt
0  2ГИС             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [48]:
industry_name = 'Разработка программного обеспечения' # наименование сферы деятельности

# текст запроса
query_5_6 = f'''select 
                    count(e.id)
                from 
                    employers e
                join 
                    employers_industries ei on e.id = ei.employer_id
                join 
                    industries i on ei.industry_id = i.id
                where 
                    i.name = '{industry_name}'
            '''

In [49]:
# результат запроса
df_emp_5_6 = pd.read_sql_query(query_5_6, connection)
df_emp_5_6

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\250393284.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [50]:
# код для получения списка городов-милионников

import requests # Импортируем модуль requests
from bs4 import BeautifulSoup # импортируем библиотеку BeautifulSoup

# определяем адрес страницы
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)  # выполняем GET-запрос

# Отправляем запрос на получение таблицы класса 'standard sortable'
page = BeautifulSoup(response.text, 'html.parser').find('table', class_='standard sortable')

cities = [] # Создаем список городов-миллионников

for td in page.find('tbody').find_all('td'): # В теле таблицы ищем тег 'a'
    if td.find('a'): # Если тег 'a' найден, получаем атрибут 'title'
        cities.append(td.find('a')['title']) # Наполняем список городов-миллионников
        
print(cities) # Проверяем список городов-миллионников

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград']


In [51]:
query_5_7_cities = '' # Создаем подзапрос с городами-милионниками

for i, city in enumerate(cities): # Перебираем в цикле номера и названия городов-милионников
    if i == 0: # Для первого города
        query_5_7_cities = "with cities as\n(\nselect " + "'" + city + "'" + " as name\nunion all \n" # Заявляем подзапрос с первым городом
    elif i < len(cities) - 1: # Для городов кроме последнего
        query_5_7_cities = query_5_7_cities + "select " + "'" + city + "'" + "\nunion all \n" # Дополняем подзапрос городом
    else: # Для последнего города
        query_5_7_cities = query_5_7_cities + "select " + "'" + city + "'" + "\n)\n" # Дополняем подзапрос городом и завершаем подзапрос

employer_name = 'Яндекс'

# Создаем запрос с присоединением таблиц areas, vacancies и employers
query_5_7_joined = query_5_7_cities \
                + f'''select 
                        c.name, count(v.id) count
                    from 
                        cities c
                    join 
                        areas a on a.name = c.name
                    join 
                        vacancies v on a.id = v.area_id
                    join 
                        employers e on e.id = v.employer_id
                    where 
                        e.name = '{employer_name}'
                    group by 
                        c.name
                '''

# Создаем запроса с добавлением строки Total
query_5_7 = "with summ as\n(\n" \
    + query_5_7_joined \
    + "\n)\n" \
    + f'''select 
            name, 
            count
        from 
            summ
        -- добавляем строку
        union all
        select 
            'Total', sum(count)
        from 
            summ
        order 
            by count
        '''



In [52]:
# результат запроса
df_area_5_7 = pd.read_sql_query(query_5_7, connection)
df_area_5_7

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\600256483.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_area_5_7 = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

## Вывод
#### Крупные компании размещают большое количество вакансий. Ряд компаний, вероятно, федерального значения, указывают своим регионом страну в целом (например, Россия), а вакансии размещают в конкретных городах, в связи с чем данные по региону компании не коррелируют с данными по регионам размещения их вакансий. Работодатели не всегда отражают сферу деятельности, выявлено около 17%, с другой тсороны, сфер деятельности может быть указано несколько. 7% работодателей занимаются разработкой программного обеспечения. Лидером по числу регионов присутствия является компания Яндекс, которая разместила вакансии в 181 регионе, в том числе от 20 вакансий в каждом городе-милионнике Российской Федерации.

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [53]:
# текст запроса
query_6_1 = f'''select 
                    count(*)
                from 
                    vacancies
                where 
                    lower(name) like '%data%' or lower(name) like '%данн%'
            '''

In [54]:
# результат запроса
df_vac_6_1 = pd.read_sql_query(query_6_1, connection)
df_vac_6_1

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\2695137618.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [55]:
# текст запроса
vacancies_selection = "(lower(name) like '%data scientist%' or lower(name) like '%data science%' \
        or lower(name) like '%исследователь данных%' or lower(name) like '%machine learning%' \
        or lower(name) like '%машинн%обучен%' or name like '%ML%') \
        and (name not like '%HTML%')"

vacancies_junior = "lower(name) like '%junior%' or experience = 'Нет опыта' or employment = 'Стажировка'"

query_6_2 = f'''select 
                    count(*)
                from 
                    vacancies
                where 
                    ({vacancies_selection}) and ({vacancies_junior})
            '''

In [56]:
# результат запроса
df_vac_6_2 = pd.read_sql_query(query_6_2, connection)
df_vac_6_2

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3529191713.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [57]:
# текст запроса
query_6_3 = f'''select 
                    count(*)
                from 
                    vacancies
                where 
                    ({vacancies_selection}) and (key_skills like '%SQL%' or key_skills like '%postgres%') 
            '''

In [58]:
# результат запроса
df_vac_6_3 = pd.read_sql_query(query_6_3, connection)
df_vac_6_3

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3050996928.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [59]:
# текст запроса
query_6_4 = f'''select 
                    count(*)
                from 
                    vacancies
                where 
                    ({vacancies_selection}) and (lower(key_skills) like '%python%')
            '''

In [60]:
# результат запроса
df_vac_6_4 = pd.read_sql_query(query_6_4, connection)
df_vac_6_4

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3875497049.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [65]:
# текст запроса
query_6_5 = f'''select
                    -- разделим навыки в массив по разделителю табуляции
                    round(avg(array_length(regexp_split_to_array(key_skills, '\t'), 1)), 2)
                from 
                    vacancies
                where 
                    ({vacancies_selection}) and key_skills is not null
            '''

In [66]:
# результат запроса
df_vac_6_5 = pd.read_sql_query(query_6_5, connection)
df_vac_6_5

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\2253810912.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [67]:
# текст запроса
query_6_6 = f'''
            select
                experience,
                -- находим среднюю зарплату, если один из концов вилки не указан, средней считаем имеющиеся данные
                round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2), 0),
                count(*)
            from 
                vacancies
            where ({vacancies_selection}) and (salary_from is not null or salary_to is not null)
                group by experience
            '''

In [68]:
# результат запроса
df_vac_6_6 = pd.read_sql_query(query_6_6, connection)
df_vac_6_6

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\127203431.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vac_6_6 = pd.read_sql_query(query_6_6, connection)


experience     round  count
0           Нет опыта   74643.0      7
1  От 1 года до 3 лет  139675.0     28
2       От 3 до 6 лет  243115.0     27

## Вывод
#### 10.5% вакансий DataScince-специалистов относятся к Junior-ам, что показывает, что людям без опыта в данной области тяжелее всего найти работу. Также часто требуемыми навыками являются Python, SQL, Postgres. К сожалению, это слишком поверхностный анализ, например обычно в требованиях указывают SQL, как замена всем типам SQL (MySQl, Postgre, MSSQL, Oracle и т.д). Реже добавляют к основному SQL навыку более точные требования к SQL или указывают опреденный подвид SQL. Но в общем анализ показывает, что знания Python, SQL являются ключевыми, и претендентам, обладающим этими навыками, легче будет найти работу. Также, что интересно, среднее кол-во навыков лежит в диапазоне от 6 до 7 навыков, т.е необходимо сделать более глубокую сегментацию, чтобы определить, сочетание каких навыков более восстребовано. Дополнительно по анализу можно сказать, что характерен достаточно резкий рост зарплаты в зависимости от опыта работы.

***

In [69]:
# найдем ТОП-5 работодателей по числу размещенных вакансий
query_self_1 = f'''select
                        count(*) as cnt,
                        employers.name
                    from 
                        vacancies 
                    join 
                        employers on vacancies.employer_id = employers.id
                    group by 
                        employers.name   
                    order by 
                        cnt desc
                    limit 5
                    '''
df_self_1 = pd.read_sql_query(query_self_1 , connection)
df_self_1

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\847135075.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_self_1 = pd.read_sql_query(query_self_1 , connection)


cnt           name
0  1933         Яндекс
1   491     Ростелеком
2   444       Тинькофф
3   428           СБЕР
4   331  Газпром нефть

#### ТОП-5 работодателей на рынке представлен телеком-компаниями, банками, специализирующимися на развитие автоматизированных технологий обслуживания и нефтяной компанией, серьезно развивающей заправочные станции самообслуживания (системы помогают пользователю без физического присутствия администраторов на месте). Это может говорить, что отрасль автоматизации активно развивается, является наиболее приоритетной и при этом имеет кадровый дефицит.

In [70]:
# Найдем среднюю зарплату в зависимости от опыта для DS вакансий и Других вакансий

vacancies_selection = "(lower(name) like '%data scientist%' or lower(name) like '%data science%' \
                        or lower(name) like '%исследователь данных%' or lower(name) like '%machine learning%' \
                        or lower(name) like '%машинн%обучен%' or name like '%ML%') \
                        and (name not like '%HTML%')"
        
query_self_2 = f'''(select
                        'Другие вакансии' as vacance_type,
                        experience,
                        -- находим среднюю зарплату, если один из концов вилки не указан, средней считаем имеющиеся данные
                        round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2), 0) as avg_salary
                    from 
                        vacancies
                    where 
                        not ({vacancies_selection}) and (salary_from is not null or salary_to is not null)
                    group by 
                        experience
                    )
                    UNION
                    (select
                        'DS вакансии',
                        experience,
                        -- находим среднюю зарплату, если один из концов вилки не указан, средней считаем имеющиеся данные
                        round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2), 0) as avg_salary
                    from 
                        vacancies
                    where 
                        ({vacancies_selection}) and (salary_from is not null or salary_to is not null)
                    group by 
                        experience
                    )
                    order by 
                        avg_salary desc
                '''

df_self_2 = pd.read_sql_query(query_self_2 , connection)
df_self_2

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\3923370371.py:37: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_self_2 = pd.read_sql_query(query_self_2 , connection)


vacance_type          experience  avg_salary
0      DS вакансии       От 3 до 6 лет    243115.0
1  Другие вакансии         Более 6 лет    166256.0
2      DS вакансии  От 1 года до 3 лет    139675.0
3  Другие вакансии       От 3 до 6 лет    132683.0
4  Другие вакансии  От 1 года до 3 лет     76403.0
5      DS вакансии           Нет опыта     74643.0
6  Другие вакансии           Нет опыта     40425.0

In [71]:
# Визуализируем результаты выборки
# функция преобразования опыта для графика
def changeSort(experience):
    if experience =='Нет опыта':
        return 1
    elif experience == "От 1 года до 3 лет":
        return 2
    elif experience == "От 3 до 6 лет":
        return 3
    elif experience == "Более 6 лет":
        return 4

data_exp_salary = df_self_2.copy() # создаем копию таблицы запроса для обработки

# Создаем колонку с данными для сортировки
data_exp_salary['Sorting'] = data_exp_salary.apply(lambda x: changeSort(x.experience), axis = 1)

# Получаем максимальное значение зарплаты, для дополнительного отображения разницы от максимума в процентах в описании колонок
max = data_exp_salary.avg_salary.max() 

# Создаем колонку с данными процентов
data_exp_salary['Percent'] = data_exp_salary.avg_salary.apply(lambda x: round(x / max, 4))
# сортируем по новому признаку Sorting
data_exp_salary.sort_values(by = ['Sorting'], inplace=True)

import plotly.express as px

fig = px.bar(
    data_exp_salary,
    x='experience',
    y='avg_salary',
    color='vacance_type', 
    color_discrete_sequence =['RGB(205,170,125)','RGB(100,149,237)'],
    hover_data = ["Percent","vacance_type"] #добавляем дополнительные поля для вывода в хинтах
)

fig.update_traces(
    hoverlabel_namelength = 0, # убираем надпись в ховере
    textposition='inside', # описание внутри
    texttemplate='   %{customdata[1]}',
    hovertemplate="".join([
        "%{customdata[1]}: ", # вытягиваем данные из переменной указаной в hover_data[]
        "Зарплата (руб.): %{y}",", ",
        "Процент: %{customdata[0]:,.2%}", # вытягиваем данные из переменной указаной в hover_data[] и форматируем чтобы проценты показывало корректно
    ]),
)

fig.update_layout(
    barmode='group',
    hovermode="x unified",
    title=dict(
        text="Зависимость средней заработной платы в рублях от опыта",
        yanchor='top',
        x=0.5,
        y=0.93
    ),
    font = dict(size=10), # уменьшим размер шрифта заголовка графика, а то не влезает
    autosize=False,# выключаем авторазмер
    width=700,# ширина
    height=700,# высота
    margin=dict(
      r=30, # отступ справа 
      b=60, # отступ снизу 
      t=100,
      l=100
    ),
    legend=dict(# параметры легенды
        borderwidth = .5, # ширина рамки в полпикселя
        bordercolor = '#cbd7de', # цвет рамки
        bgcolor = '#f2faff', # цвет фона
        orientation="h", # показать сверху
        yanchor="bottom", # вертикальное выравнивание по нижней части легенды относительно холста
        y=1.01, # сдвигаем на 1% выше, чтобы не сливалась нижняя граница рамки легенды с графиком
        xanchor="right", # выравнивание блока легенды по правому краю
        x=1, # никаких сдвигов не надо, будет на уровне с графиком по правому краю
        title_text="", # убираем title для легенды
    ),      
    yaxis=dict(
        title_text="Средняя зарплата (руб.)", # меняем название заголовка оси абсцисс
        titlefont=dict(size=15),  # ее размер 
        range = (0,30e4), # увеличивем диапазон отражения по оси ординат, чтобы не срезалась отрисовка процентов над левым баром
    ),
    xaxis=dict(
        title_text="Опыт", # меняем название заголовка оси абсцисс
        titlefont=dict(size=15), # ее размер 
    ), 
    paper_bgcolor="#eaf7ff", # цвет фона холста
)

fig.show()

#### Разделение по зависимости Опыта к зарплате между DS вакансиями и Другими, показывает, что в любых сегментах Опыта (кроме более 6 лет, там вообще нет данных по DS), Зарплата DS специалиста почти в 2 раза выше, чем любой другой специальности. Отсутствие Опыта более 6 лет, говорит, что специальность для HH.RU достаточно молодая на рынке труда. Также видна "крутая" зависимость роста зарплаты от опыта, что для DS, что для других специальностей. Средняя зарплата DS вообще без опыта, составляет 74к, что является соразмерной средней зарплатой для других специальностей с опытом 1 - 3 года, что несколько странно. Это говорит о том, что вообще нет смысла устраиваться на работу по другой специальности и набирать опыт, когда сразу можно можно попытаться протиснуться без опыта на такие же денежные условия.

In [72]:
# Анализ самых распространенных навыков вакансий DS

vacancies_selection = "(lower(name) like '%data scientist%' or lower(name) like '%data science%' \
                        or lower(name) like '%исследователь данных%' or lower(name) like '%machine learning%' \
                        or lower(name) like '%машинн%обучен%' or name like '%ML%') \
                        and (name not like '%HTML%')"
                        
query_self_3 = f'''select
                        -- разделим навыки по разделителю табуляции
                        regexp_split_to_table(key_skills, '\t') skill,
                        count(*) as cnt,
                        -- посчитаем процент количества каждого из навыков
                        round(count(*) * 100.0 / count(count(*)) over(),2) percent
                    from 
                        vacancies
                    where 
                        {vacancies_selection}
                    group by skill
                    order by cnt desc
                    limit 20
                '''

df_self_3 = pd.read_sql_query(query_self_3 , connection)
df_self_3

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\238655178.py:23: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



skill  cnt  percent
0                         Python  348    82.46
1                            SQL  191    45.26
2               Machine Learning  114    27.01
3                            Git   66    15.64
4      Математическая статистика   62    14.69
5                  Data Analysis   54    12.80
6                   Data Science   52    12.32
7                         Pandas   52    12.32
8                             ML   49    11.61
9                Английский язык   48    11.37
10                         Linux   47    11.14
11                 Анализ данных   46    10.90
12                       PyTorch   46    10.90
13                      Big Data   42     9.95
14  Математическое моделирование   41     9.72
15         Статистический анализ   40     9.48
16                           NLP   40     9.48
17                           C++   39     9.24
18                         Spark   38     9.00
19                    Tensorflow   35     8.29

#### Из запроса видно:

* В 81% одним из навыков в вакансии DS является Python
* На втором месте идет SQL с 45%
* Machine learning и ML вместе занимают около 38%
* Далее следует Анализ данных и Data Analysis около 24%

#### Таким образом, обладая вышеуказанными навыками, проще найти работу по профессии DS

In [73]:
# Анализ самых распространенных городов для вакансий DS

vacancies_selection_v = "(lower(v.name) like '%data scientist%' or lower(v.name) like '%data science%' \
                        or lower(v.name) like '%исследователь данных%' or lower(v.name) like '%machine learning%' \
                        or lower(v.name) like '%машинн%обучен%' or v.name like '%ML%') \
                        and (v.name not like '%HTML%')"
                        
query_self_3 = f'''select
                        a.name,
                        count(v.id)
                    from 
                        areas a
                    join vacancies v on a.id = v.area_id
                    where 
                        {vacancies_selection_v}
                    group by a.name
                    order by 2 desc
                    limit 10
                '''

df_self_3 = pd.read_sql_query(query_self_3 , connection)
df_self_3

C:\Users\Sl1mka\AppData\Local\Temp\ipykernel_1524\1869548182.py:21: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0           Москва    217
1  Санкт-Петербург     64
2      Новосибирск     23
3  Нижний Новгород     20
4           Казань     16
5           Алматы     16
6            Минск      9
7            Томск      8
8     Екатеринбург      7
9          Армения      6

#### Из запроса видим ТОП-5 городов для поиска работы датасаентистам. Также видим географический разброс для професии:

* Лидерами являются города федерального значения Москва и Санкт-Петербург;
* Работа для датасаентистов есть и с удалением на восток страны: 
    * в Приволжском федеральном округе - Нижний Новгород, Казань;
    * в Уральском федеральном округе - Екатеринбург;
    * в Сибирском федеральном округе - Новосибирск, Томск;
* В выборку попали вакансии в двух столицах странг СНГ;
* На последнюю позицию ТОП-10 попала Армения, как страна в целом, а не город, из чего можно предположить хорошую возможность удаленной работы в ней для датасаентистов

In [75]:
connection.close()

# Общий вывод по проекту
### На рынке находится около 50 тысяч открытых вакансий от 23т работодателей. Больше всего вакансих предлагают гиганты рынка, например Яндекс, Ростелеком, Тинькофф, СБЕР, Газпромнефть. Это говорит либо о расширении компаний, либо о большой текучке кадров.

### Для большинства требуется от 1 до 3 лет опыта работы (около 25 тыс.), далее по популярности от 3 до 6 лет. Наимение популярными являются специалисты с более 6 годами опыта. Таким образом, слежнее всего на рынке труда приходится самым молодым и самым опытным специалистам. Можно сделать вывод, что работодатели ищут специалистов с опыт, но не слишком дорогих. В базе представленны вакансии из 1300 населенных пунктов. Географически, вакансии сконцетрированы в столицах и крупных городах, а именно в Москве, СПб, Минске, Новосибирске. Болтшинство работодателей не побликуют информацию о заработной плате, а те, кто предоставляют такие сведения, в среднем предлогают от 71 до 110 т.р.. Больше всего вакансий требуют полной занятосни с полным рабочем днем или удаленной работой.